### ISSUE TITLE
Old deploys take up way heaps of space

### ISSUE THREAD

**Issue by dacook** [openfoodfoundation/ofn-install#899]
## Description
We store the last ten deploys on a server in the `releases-old` folder, but this can take up lots of space. The 5 oldest are gzipped but still take up space.

The majority of the space used is from `node_modules`, currently 260mb (therefore 2.6gb used).
Logs also take up a bit, but this is minimised when zipped.

## Expected Behavior
There shouldn't be too much space used up

## Actual Behaviour

```
# au-prod
du -sh apps/openfoodnetwork/releases-old
5.5G	apps/openfoodnetwork/releases-old
```

## Possible Fix
 Perhaps `node_modules` could be linked to a single shared folder (I think assets might be set up like this?).

But I think we can simply just store less old releases. I don't think we ever really refer to them.

**Comment by mkllnk** [comment]
When I go back to old releases then it's mostly for log files to see what happened two weeks ago. So we could think of storing them centrally with logrotate.

Keeping node modules centrally has the advantage of a quicker installation process because we don't need to duplicate all of them. But pruning is a challenge. If we just prune according to the current deploy then we lose the backup of the old releases. Or are the installation packages cached independently?

**Comment by dacook** [comment]
Good point about logs, we should consider storing them centrally.

For node_modules, I'm not sure if there's a global local cache (that would be a good idea). I had assumed it was copied from the previous release but I now see that's not the case. If it's not cached we should do something about that to speed up deploys.

But in any case, I reckon just reducing the  number of old releases is simple and would be ok. How about the last 4 releases (that should cover almost a month of logs)?

**Comment by mkllnk** [comment]
> How about the last 4 releases (that should cover almost a month of logs)?

Yes, good idea. We may do more deploys when something goes wrong and then we lose logs but that's a different issue.

> For node_modules, I'm not sure if there's a global local cache

We could also copy the previous release, install and prune if we don't have another cache.

**Comment by dacook** [comment]
Silly me I mustn't have been looking closely when I first wrote this. There are ten things in the `releases-old` dir, but half are DB dumps. There's only 5 old releases (plus the current one). Eg:
```bash
ofn-admin@prod3:/home/openfoodnetwork/apps/openfoodnetwork/releases-old$ ls -1tr
2024-06-10-231508
2024-06-10-231508.sql.gz
2024-06-18-003347
2024-06-18-003347.sql.gz
2024-06-19-055150
2024-06-19-055150.sql.gz
2024-06-25-051021
2024-06-25-051021.sql.gz
2024-07-02-043842
2024-07-02-043842.sql.gz
2024-07-09-043115  # <-- current release
```

The logic is here, and takes a bit of explanation..
https://github.com/openfoodfoundation/ofn-install/blob/1e669a3f0d8973eb99758f7558511b336642ac29/roles/deploy/tasks/rollbacks.yml#L52
1. `set -o pipefail`: Fail on any errors in the pipeline
2. `ls -1tr`: List each file on **1** line, in order of **t**ime updated, **r**eversed.
3. `head -n -10`: Strip the last 10 lines.

This deletes the one oldest file/dir
```
ofn-admin@prod3:/home/openfoodnetwork/apps/openfoodnetwork/releases-old$ ls -1tr | head -n-10
2024-06-10-231508
```
This doesn't seem quite right, I think it means we delete the SQL or dir on alternating weeks. So it keeps about 4.5 old releases, not five. We should change that number to 9. But it will only make a small difference on disk usage...

**Comment by dacook** [comment]
So I think the focus should be on `node_modules`, but we could also consider the `tmp` folder too. Eg:

```
  273.5 MiB [##########] /node_modules
  129.7 MiB [####      ] /.git
  119.5 MiB [####      ] /tmp
  117.8 MiB [####      ] /log
   45.1 MiB [#         ] /public
...
```

I think re-using `/tmp` and `public` would be advantageous, but needs to be managed properly, which is hard. Better to avoid it (I also considered this for CI).

For `node_modules`, after a bit of reading, I found that:
* yarn keeps a global cache storage of all downloaded packages, so it doesn't re-download them every time
* It does copy the packages each time, but there's a setting you can change to hardlink from the cache them instead: https://yarnpkg.com/configuration/yarnrc#nmMode
   * The feature is discussed a lot [here](https://github.com/yarnpkg/yarn/issues/1064). **I think it would be worth using this.** But we should test it to ensure that packages aren't deleted when old releases are deleted.
